### PR TITLE
Expose router parameters to JS

### PR DESCRIPTION
**Pull request by thisandagain**

### CODE REVIEW

File: app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java [hunk 1]
```diff
@@ -7,21 +7,24 @@
...
     private Context mContext;
     private SharedPreferences mPrefs;
+    private JSONObject mRoute;
     private String mPrefKey;
     private String mPageState;
 
     public static final String WEBMAKER_PREFS = "WEBMAKER";
 
-    public WebAppInterface(Context c) {
+    public WebAppInterface(Context c, JSONObject routeParams) {
```
Comment (Pomax) [inline app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java]: we're still in Java, let's make a fall-through constructor:
```
public WebAppInterface(Context c) {
  this(c, null);
}
public WebAppInterface(Context c, JSONObject routeParams) {
  ...
}
```

File: app/src/main/java/mozilla/org/webmaker/fragment/WebviewFragment.java [hunk 2]
```diff
@@ -49,7 +49,7 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
         }
 
         // Create webview
-        mWebView = new WebmakerWebView(mView.getContext(), pageId);
+        mWebView = new WebmakerWebView(mView.getContext(), pageId, null);
```
Comment (Pomax) [inline app/src/main/java/mozilla/org/webmaker/fragment/WebviewFragment.java]: no need for the explicit `null` if we make sure we have a fall-through constructor for `WebmakerWebView`

File: app/src/main/java/mozilla/org/webmaker/view/WebmakerWebView.java [hunk 3]
```diff
@@ -7,18 +7,19 @@
 
 import mozilla.org.webmaker.client.WebClient;
 import mozilla.org.webmaker.javascript.WebAppInterface;
+import org.json.JSONObject;
 
 public class WebmakerWebView extends WebView {
 
     @SuppressLint("SetJavaScriptEnabled")
-    public WebmakerWebView(Context context, String pageName) {
+    public WebmakerWebView(Context context, String pageName, JSONObject routeParams) {
```
Comment (Pomax) [inline app/src/main/java/mozilla/org/webmaker/view/WebmakerWebView.java]: fall-throughs :heart:
```
public WebmakerWebView(Context context, String pageName) {
  this(context, pageName, null);
}

public WebmakerWebView(Context context, String pageName, JSONObject routeParams) {
  ...
}
```

File: www_src/pages/map/map.jsx [hunk 4]
```diff
@@ -29,6 +29,13 @@ var Map = React.createClass({
     var height = 380;
     var gutter = 20;
 
+    if (window.Android) {
```
Comment (Pomax) [inline www_src/pages/map/map.jsx]: do we want to keep this in for the PR?

Comment (thisandagain) [inline www_src/pages/map/map.jsx]: Nah. I'll pull in the final and just add a note in the README.md

Comment (Pomax) [inline www_src/pages/map/map.jsx]: \o/

Comment (k88hudson) [inline www_src/pages/map/map.jsx]: I was thinking it's probably worth replacing the binding mixin stuff with this instead, so we get `getRouteParams` as a method in that mixin

Comment (k88hudson) [inline www_src/pages/map/map.jsx]: happy to do this in another patch though after you land this

**Comment by thisandagain** [comment]
Re-R @k88hudson and @Pomax

**Comment by thisandagain** [comment]
Gah! Sorry @potch :cat:

File: app/src/main/java/mozilla/org/webmaker/activity/Map.java [hunk 5]
```diff
@@ -20,5 +24,4 @@ public boolean onCreateOptionsMenu(Menu menu) {
     public boolean onOptionsItemSelected(MenuItem item) {
         return item.getItemId() == R.id.action_settings;
     }
-}
-
+}
```
Comment (Pomax) [inline app/src/main/java/mozilla/org/webmaker/activity/Map.java]: ohnoes O_O

File: app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java [hunk 6]
```diff
@@ -59,4 +66,10 @@ public void logText(String txt){
     public void setView(String url) {
         Router.sharedRouter().open(url);
     }
+
+    @JavascriptInterface
+    public String getRouteParams() {
+        Log.v("Router", mRoute.toString());
+        return mRoute.toString();
```
Comment (Pomax) [inline app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java]: I suspect this'll throw nullpointer exceptions at us at some time in the future when we forgot about this already again...
```
if (mRoute==null) {
  return "";
}
return mRoute.toString();
```

File: app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java [hunk 7]
```diff
@@ -59,4 +66,10 @@ public void logText(String txt){
...
+    }
 }
```
Comment (Pomax) [inline app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java]: newline

File: app/src/main/java/mozilla/org/webmaker/activity/Map.java (hunk 5 above)
Comment (thisandagain) [inline app/src/main/java/mozilla/org/webmaker/activity/Map.java]: Wat. *Shakes fist at Android Studio*

File: www_src/lib/router.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+module.exports = {
+  getRouteParams: function () {
+    var p = {};
+    if (window.Android) {
+      p = JSON.parse(window.Android.getRouteParams());
+    }
+
```
Comment (Pomax) [inline www_src/lib/router.js]: tiny nit: remove empty line?

File: app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java (hunk 6 above)
Comment (k88hudson) [inline app/src/main/java/mozilla/org/webmaker/javascript/WebAppInterface.java]: hm, if you don't want to throw exceptions from JSON.parse, if it's possible this is `null`, you should alternatively return "{}".

**Comment by k88hudson** [comment]
Looks good to me.

**Comment by Pomax** [comment]
I cannea find na thing ta fault. R+
